PaintingTrivia.py

```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
-        key = 'right_guesses' if chosen_painter == session[
-            'selected_painter_id'] else 'wrong_guesses'
-        session[key] += 1
+            print "Something failed"
 
-    #http://docs.sqlalchemy.org/en/latest/orm/tutorial.html#using-exists
-    #the instructions below correspond to the following sql statement:
-    # SELECT painter.id AS painter_id, painter.name AS painter_name FROM painter
-    # WHERE EXISTS (SELECT * FROM painting WHERE painter.id = painting.painter_id)
-    # ORDER BY random() LIMIT 4
+    if session['quiz'] is None:
+        # Obtain the painters and the paintings randomly
+        selected_painters, selected_paintings = initialize_images(num_painters=10)
+        quiz = PainterQuiz(selected_paintings, selected_painters)
+        session['quiz'] = pickle.dumps(quiz)
+    else:
+        quiz = pickle.loads(session['quiz'])
 
-    stmt = exists().where(Painter.id == Painting.painter_id)
-    painters = Painter.query.filter(stmt).order_by(func.random()).limit(4).all()
-    selected_painter = random.choice(painters)
-    selected_painting = Painting.query.filter(Painting.painter == selected_painter).order_by(func.random()).limit(
-        1).first()
+    # If all of the images have been seen, show again, the incorrect ones
+    question = quiz.next_question()
 
-    session['selected_painter_id'] = selected_painting.painter.id
+    session['quiz'] = pickle.dumps(quiz)
+    if not question:
+        return redirect(url_for('show_quiz_results'))
     return render_template('guess_the_painter.html',
-                           painters=painters,
-                           selected_painter=selected_painter,
-                           selected_painting=selected_painting,
-                           right_guesses=session['right_guesses'],
-                           wrong_guesses=session['wrong_guesses'],
+                           painters=question.option_list,
+                           selected_painter=question.correct_option,
+                           selected_painting=question.element.values()[0],
+                           right_guesses=sum(quiz.score),
+                           wrong_guesses=len(quiz.score) - sum(quiz.score),
                            username=session['username'])
 
 
+def initialize_images(num_painters=10):
+    stmt = exists().where(Painter.id == Painting.painter_id)
+    selected_painters = Painter.query.filter(stmt).order_by(func.random()).limit(num_painters).all()
+
+    painters_dict = dict()
+    paintings_dict = dict()
+    selected_paintings = []
+    for key, painter in enumerate(selected_painters):
+        paintings_aux = Painting.query.filter(Painting.painter_id == painter.id).order_by(func.random()).limit(1).all()
+        selected_paintings.extend(paintings_aux)
+        painters_dict[key] = painter
+        paintings_dict[key] = paintings_aux[0]  # get the first
+    return painters_dict, paintings_dict
+
+
+def initialize_saint_images(num_saints=10):
```

Comment:
This function is almost one-to-one copy of the previous one. please refactor to have only one